### PR TITLE
Fix longformer onnx broken export

### PR DESCRIPTION
**Pull request by fxmarty**
This PR fixes the ONNX export of longformer, that was **silently** broken for several cases:

* the export registers `padding_len > 0` as a constant equal to `True`, hence during inference in the dynamic case `padding_len == 0`, we would still go through the path `padding_len > 0` that would then contain negative indexing making some ONNX nodes fail (gather). This PR fixes the negative indexes.
* the export registers `hidden_states.size(1) == window_overlap * 2:` as a constant equal `True` during the export, hence using the converted ONNX model was failing when the `input_ids` length was strictly greater than `attention_window` (case where the `else` path should be taken). This PR removes the path `hidden_states.size(1) == window_overlap * 2:`, since the other path can handle this case as well.

Had to run `make fix-copies` than modified led model as well.

@michaelbenayoun @lewisbails Where should I add tests for this? Optimum?

## Before submitting
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by michaelbenayoun** [comment]
Yes, I would add the support for Longformer in `optimum`, and the tests there as well.

File: src/transformers/models/longformer/configuration_longformer.py [hunk 1]
```diff
@@ -221,7 +221,10 @@ def generate_dummy_inputs(
         )
         import torch
 
+        # for some reason, replacing this code by inputs["global_attention_mask"] = torch.randint(2, inputs["input_ids"].shape, dtype=torch.int64)
+        # makes the export fail randomly
```
Comment (michaelbenayoun) [inline src/transformers/models/longformer/configuration_longformer.py:225]: It might try to attend to some forbidden positions?

**Review by michaelbenayoun — APPROVED** [review]
LGTM (when the tests pass)

File: src/transformers/models/led/modeling_led.py [hunk 2]
```diff
@@ -411,29 +411,26 @@ def _chunk(hidden_states, window_overlap, onnx_export=False):
             return hidden_states.as_strided(size=chunk_size, stride=chunk_stride)
 
         # When exporting to ONNX, use this separate logic
```
Comment (lewtun) [inline src/transformers/models/led/modeling_led.py:413]: Wondering if it would make sense to include this in an `else` statement to make the code a bit clearer to read?

File: src/transformers/models/longformer/modeling_longformer.py [hunk 3]
```diff
@@ -782,29 +782,26 @@ def _chunk(hidden_states, window_overlap, onnx_export=False):
             return hidden_states.as_strided(size=chunk_size, stride=chunk_stride)
 
         # When exporting to ONNX, use this separate logic
```
Comment (lewtun) [inline src/transformers/models/longformer/modeling_longformer.py:784]: Similar comment here about using an `else` statement

File: src/transformers/models/longformer/modeling_longformer.py [hunk 4]
```diff
@@ -1294,7 +1291,7 @@ def forward(
 
         is_index_masked = attention_mask < 0
         is_index_global_attn = attention_mask > 0
-        is_global_attn = is_index_global_attn.flatten().any().item()
+        is_global_attn = torch.any(is_index_global_attn).item()  # the ONNX export should record is_global_attn == True
```
Comment (lewtun) [inline src/transformers/models/longformer/modeling_longformer.py:1294]:
```suggestion
         # Record `is_global_attn == True` to enable ONNX export
        is_global_attn = torch.any(is_index_global_attn).item()
```

**Review by lewtun — APPROVED** [review]
Thanks a lot for fixing this sneaky bug @fxmarty !

The modeling changes look good to me, but gently pinging @ydshieh for his opinion too.

File: src/transformers/models/longformer/modeling_longformer.py (hunk 4 above)
Comment (ydshieh) [inline src/transformers/models/longformer/modeling_longformer.py:1294]: Will try to review in more details later, but a question here:

Why we need to :
 - use `torch.any` instead of the tensor's `any()` method
 - remove `flatten`

Are these just some simplification rather than necessarity?

Furthermore, could you explain a bit to me (ONNX newbie) what it means 

``` 
# Record `is_global_attn == True` to enable ONNX export
```

File: src/transformers/models/led/modeling_led.py [hunk 5]
```diff
@@ -411,29 +411,26 @@ def _chunk(hidden_states, window_overlap, onnx_export=False):
...
-
-            # TODO replace this with
-            # > return hidden_states.unfold(dimension=1, size=window_overlap * 2, step=window_overlap).transpose(2, 3)
-            # once `unfold` is supported
+        # have to use slow implementation since as_strided, unfold and 2d-tensor indexing aren't supported (yet) in ONNX export
+
+        # TODO replace this with
+        # > return hidden_states.unfold(dimension=1, size=window_overlap * 2, step=window_overlap).transpose(2, 3)
+        # once `unfold` is supported
+        # the case hidden_states.size(1) == window_overlap * 2 can also simply return hidden_states.unsqueeze(1), but that's control flow
```
Comment (ydshieh) [inline src/transformers/models/led/modeling_led.py:419]: So ONNX export can't never work correctly when using tensors in the control flow, if we have different tensor properties between the export time and the inference time?

File: src/transformers/models/longformer/modeling_longformer.py [hunk 6]
```diff
@@ -1350,14 +1347,15 @@ def custom_forward(*inputs):
             all_hidden_states = all_hidden_states + (hidden_states,)
 
         # undo padding
+        # this path should be recorded in the ONNX export, it is fine with padding_len == 0 as well
         if padding_len > 0:
```
Comment (ydshieh) [inline src/transformers/models/longformer/modeling_longformer.py:1351]: Any reason not to remove this condition `if padding_len > 0:`?

File: src/transformers/models/led/modeling_led.py (hunk 5 above)
Comment (fxmarty) [reply src/transformers/models/led/modeling_led.py:419]: Basically onnx export follows the warning here: https://pytorch.org/docs/stable/generated/torch.jit.trace.html

If there is control flow dependent on things such as batch size / sequence length, yes indeed, the exported model will have failing cases. In case there are static attributes and only a path is recorded. For example we never enter this control flow during the export and it's totally fine: https://github.com/huggingface/transformers/blob/v4.15.0/src/transformers/models/bert/modeling_bert.py#L229

Currently the tests are inexistent in this regard: the export is done with sequence length = 8, the validation test is done with sequence length = 9. Hence the silent errors.

File: src/transformers/models/longformer/modeling_longformer.py (hunk 4 above)
Comment (fxmarty) [reply src/transformers/models/longformer/modeling_longformer.py:1294]: Right, it's probably unnecessary. I'll fix.

The comment means that in a normal case scenario, the control flows should be done having `is_global_attn == True`. This means that the example provided to `torch.onnx.export` should have at least at least one id of `global_attention_mask` non-zero.

Comment (ydshieh) [reply src/transformers/models/longformer/modeling_longformer.py:1294]: Thank you. I understand the comment better now.

One more question however, if the case `is_global_attn = True` is recorded and used few lines below (note here is an assignment, not `==`) during export time, would we have trouble during inference time if we feed an input without any positive global attention mask?

File: src/transformers/models/longformer/modeling_longformer.py (hunk 6 above)
Comment (ydshieh) [reply src/transformers/models/longformer/modeling_longformer.py:1351]: I think we can remove this `if` statement, and keep your change `hidden_states.shape[1] - padding_len` below.
This might be clear. (unless there is some optimization consideration, i.e. we don't want to enter this block to save some running time)

**Review by ydshieh — APPROVED** [review]
Thank you @fxmarty!

File: src/transformers/models/led/modeling_led.py (hunk 2 above)
Comment (fxmarty) [reply src/transformers/models/led/modeling_led.py:413]: Personally agree it's more readable. I will leave it as is following https://github.com/huggingface/transformers/pull/17176#discussion_r915711144

File: src/transformers/models/longformer/modeling_longformer.py (hunk 4 above)
Comment (fxmarty) [reply src/transformers/models/longformer/modeling_longformer.py:1294]: Yes exactly. IMO the ONNX export is lacking in this sense: since we record only a single path, there may well be cases where the exported ONNX is invalid / will throw weird errors), while it is valid in other cases.

I was thinking to have some kind of support (in Optimum exporters) for an external file to the `.onnx` that would specify which cases are supported with the exported onnx.

File: src/transformers/models/longformer/modeling_longformer.py (hunk 6 above)
Comment (fxmarty) [reply src/transformers/models/longformer/modeling_longformer.py:1351]: Yes I left it thinking the case `padding_len == 0` would be faster if we don't go in this if code. But I'll remove the condition, I agree it's not clear currently.

**Comment by lewtun** [comment]
Thanks for iterating @fxmarty ! 

Since @ydshieh has also approved, gently pinging @sgugger for final approval :)

**Review by sgugger — APPROVED** [review]
Thanks for the fix!

**Comment by adithya1111** [comment]
Hello @sgugger @lewtun  : I am stuck with the same issue. Thanks for the fix. This may be a noob question. But wanted to check when would this change be reflected in the PyPi package ? Is it during the next release ? If so do we know when would that be happening ?

**Comment by ydshieh** [comment]
@adithya1111 Somewhere in the next week if I remember correctly.

**Comment by fxmarty** [comment]
Hello @adithya1111 yes this will be in the next release, feel free to try the main branch meanwhile. In any case, I would advise you to be very careful with the exported ONNX model, and to check that the outputs are on par with PyTorch for your target sequence length. You can possibly edit the ONNX export code if you want to use the exported ONNX with a different sequence length, as explained in my messages above.

For reference: https://github.com/huggingface/optimum/issues/503

**Comment by adithya1111** [comment]
Thanks a lot for the comments . Another question. I used the default parameters when training. So would that mean the Global Attention Mask = 0 . I see that we are now setting the global_attention_mask[:, ::2] = 1 . Could this lead to a discrepancy. 

My original models results are `logits=tensor([[[ 0.6209,  0.0719,  0.1107, -0.5316],
         [ 3.0321, -0.2787, -0.6460, -2.5359],
         [ 2.6904,  0.1169, -0.7495, -2.8346],
         [ 0.6474,  0.0761,  0.1041, -0.5438]]]`

And my ONNX converted predictions are `[array([[[ 0.49600145,  0.08062335,  0.12902021, -0.4010917 ],
         [ 3.0400352 , -0.34643874, -0.6276542 , -2.444679  ],
         [ 2.158992  ,  0.02124629, -0.5462518 , -2.094074  ],
         [ 0.6290194 ,  0.06919068,  0.10753635, -0.5197539 ]]],
       dtype=float32)]`

Its close but there are some discrepancies

**Comment by fxmarty** [comment]
@adithya1111 Could you open an issue in https://github.com/huggingface/optimum/issues with a reproducible code? ONNX export through `transformers.onnx` will eventually depend on `optimum.exporters` so we can track the issue there.

Thanks!

**Comment by adithya1111** [comment]
Created a new issue. Thanks

**Comment by adithya1111** [comment]
@sgugger @fxmarty @ydshieh @lewtun : Does the latest release fix this issue ?

**Comment by fxmarty** [comment]
With the release you will have no error at the export. However, following the discussion above (see the closed comments), and as well in https://github.com/huggingface/transformers/issues/20275 , https://github.com/huggingface/optimum/issues/503 , https://github.com/huggingface/optimum/issues/505 , you can expect to have non-meaningful output running the ONNX with sensibly different sequence length than the example provided to `torch.onnx.export` during the conversion.

This is WIP to add options to customize more the export, refer to https://github.com/huggingface/optimum/pull/522
